lib/api/model/model.dart

```diff
@@ -698,3 +710,74 @@ class DmMessage extends Message {
   @override
   Map<String, dynamic> toJson() => _$DmMessageToJson(this);
 }
+
+enum MessageEditState {
+  none,
+  edited,
+  moved;
+
+  // Adapted from the shared code:
+  //   https://github.com/zulip/zulip/blob/1fac99733/web/shared/src/resolved_topic.ts
+  // Pattern for an arbitrary resolved-topic prefix.
+  // These always begin with the canonical prefix, but can go on longer.
+  // It's designed to remove a weird "✔ ✔✔ " prefix, if present.
+  static RegExp _resolvedTopicPrefixRe = RegExp('^✔ [ ✔]*');
+
+  /// Whether two topics are equal, ignoring any resolved-topic prefix.
+  ///
+  /// When a topic is resolved, the clients agree on adding a ✔ prefix to the
+  /// topic string. Topics whose only difference is the ✔ prefix are considered
+  /// the same. This helper can be helpful when checking if a message has been
+  /// moved.
+  static bool areSameTopic(String topic, String prevTopic) {
+    // TODO(#744) Extract this to its own home to support "mark as resolve".
+    topic = topic.replaceFirst(_resolvedTopicPrefixRe, '');
+    prevTopic = prevTopic.replaceFirst(_resolvedTopicPrefixRe, '');
+
+    return topic == prevTopic;
+  }
+
+  static MessageEditState _readFromMessage(Map<dynamic, dynamic> json, String key) {
+    // Adapted from the web app:
+    //   https://github.com/zulip/zulip/blob/c31cebbf68a93927d41e9947427c2dd4d46503e3/web/src/message_list_view.js#L68-L118
```

Comment:
```suggestion
    // Adapted from `analyze_edit_history` in the web app:
    //   https://github.com/zulip/zulip/blob/c31cebbf68a93927d41e9947427c2dd4d46503e3/web/src/message_list_view.js#L68-L118
```

The function name isn't strictly needed — the permalink, complete with line numbers, carries all the information and is the most important thing — but is a helpful signpost for a reader who hasn't followed the link.